### PR TITLE
Chapter 5 - Team Fight

### PR DESCRIPTION
**Pull request by RobinKa**
- We start outside the rosh pit
- Enemies are around top rune location
- Pan camera to them and back
- Start teamfight
- When teamfight is over, give player a tp
- Player attempts to use it, gets baited into using voice chat again
- Finally tp player to fountain

Will cause some conflicts with other CH5 sections.

Mostly complete, missing things:
- Handle dialog for dying (this is practically impossible to happen, so I didn't bother yet)
- More teamfight logic (right now its just autoattacking lvl 25 heroes)
- Maybe we'll want to change the heroes

image

### CODE REVIEW

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    graph = tg.withGoals(_ => goalTracker.getGoals(),
+        tg.seq([
+            shared.spawnFriendlyHeroes(shared.outsidePitLocation),
+            shared.spawnEnemyHeroes(shared.enemyLocation),
+
+            // Pan to enemies and back while dialog is playing
+            tg.immediate(_ => GameRules.SetTimeOfDay(0.5)),
+            tg.immediate(_ => goalSpotEnemies.start()),
+            tg.immediate(_ => freezePlayerHero(true)),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:61]: Combine those 3 immediates into one.

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                    tg.wait(0.75),
+                    tg.immediate(_ => goalKillEnemyHeroes.start()),
+                    tg.wait(0.75),
+                    tg.panCameraExponential(_ => getPlayerCameraLocation(), _ => playerHero.GetAbsOrigin(), 2),
+                ])
+            ]),
+            tg.immediate(_ => freezePlayerHero(false)),
+
+            // Start teamfight
+            tg.immediate(ctx => shared.getLivingHeroes(ctx).forEach(hero => {
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:77]: Combine immediates

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                    OrderType: UnitOrder.ATTACK_MOVE,
+                    Position: (hero.GetTeam() === DotaTeam.GOODGUYS ? shared.enemyLocation : shared.outsidePitLocation).__add(RandomVector(300)),
+                    UnitIndex: hero.entindex(),
+                })
+            })),
+
+            // Wait for every enemy to die
+            // TODO: Add logic for if the player dies (although not a disaster atm. either since they just respawn)
+            tg.loop(ctx => shared.getLivingEnemyHeroes(ctx).length > 0, ctx => tg.seq([
+                tg.immediate(_ => goalKillEnemyHeroes.setValue(5 - shared.getLivingEnemyHeroes(ctx).length)),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:88]: Change `5` to actual length of the enemy heroes array.

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                tg.wait(1),
+            ])),
+            tg.immediate(_ => goalKillEnemyHeroes.setValue(5)),
+            tg.immediate(_ => goalKillEnemyHeroes.complete()),
+
+            // Play win dialog
+            tg.immediate(ctx => shared.getLivingFriendlyHeroes(ctx).forEach(hero => hero.StartGesture(GameActivity.DOTA_VICTORY))),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:95]: Combine immediates.

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                    tg.completeOnCheck(_ => !voicePressed, 0.1),
+                    tg.audioDialog(LocalizationKey.Script_5_5v5_13, LocalizationKey.Script_5_5v5_13, ctx => ctx[CustomNpcKeys.SunsFanMudGolem]),
+                ])
+            ]),
+            tg.immediate(_ => goalPromiseCarryTp.complete()),
+
+            // Actually tp the player to the fountain
+            tg.immediate(_ => freezePlayerHero(true)),
+            tg.immediate(_ => playerHero.RemoveItem(getOrError(playerHero.FindItemInInventory("item_tpscroll"), "Could not find tp scroll"))),
+            tg.immediate(_ => playerHero.SetAbsOrigin(radiantFountainLocation)),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:132]: Combine

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            tg.panCameraExponential(_ => getPlayerCameraLocation(), radiantFountainLocation, 2),
+            tg.immediate(ctx => shared.disposeHeroes(ctx)),
+            tg.immediate(_ => freezePlayerHero(false)),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:135]: Combine

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                    tg.wait(6),
+                    tg.completeOnCheck(_ => !voicePressed, 0.1),
+                    tg.audioDialog(LocalizationKey.Script_5_5v5_13, LocalizationKey.Script_5_5v5_13, ctx => ctx[CustomNpcKeys.SunsFanMudGolem]),
+                ])
+            ]),
+            tg.immediate(_ => goalPromiseCarryTp.complete()),
+
+            // Actually tp the player to the fountain
+            tg.immediate(_ => freezePlayerHero(true)),
+            tg.immediate(_ => playerHero.RemoveItem(getOrError(playerHero.FindItemInInventory("item_tpscroll"), "Could not find tp scroll"))),
```
Comment (Shushishtok) [inline game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:131]: I'm not entirely sold on the fake TP thing, let's make them actually TP.
If needed, maybe rearrange the script so you first promise to always have a TP, then TP? Not jebaits. Might be better.

**Review by Shushishtok — CHANGES_REQUESTED** [review]

File: game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts (hunk 7 above)
Comment (RobinKa) [reply game/scripts/vscripts/Sections/Chapter5/SectionTeamFight.ts:132]: Promise to do that later, just wanted to get something functioning up! Also jebait is too fun to remove.

**Pull request merged**